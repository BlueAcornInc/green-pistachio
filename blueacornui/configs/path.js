/**
 * Copyright Â© 2016 Magento. All rights reserved.
 * See COPYING.txt for license details.
 */

'use strict';

/**
 * Define paths.
 */

var webroot = '../webroot/';

module.exports = {
    binMage: webroot + 'bin/magento',
    pub: webroot + 'pub/static/',
    code: webroot + 'app/code/',
    design: webroot + 'app/design/',
    tmpLess: webroot + 'var/view_preprocessed/less/',
    tmpSource: webroot + 'var/view_preprocessed/source/',
    tmp: webroot + 'var',
    css: {
        setup: webroot + 'setup/pub/styles',
        updater: webroot + '../magento2-updater/pub/css'
    },
    less: {
        setup: webroot + 'app/design/adminhtml/Magento/backend/web/app/setup/styles/less',
        updater: webroot + 'app/design/adminhtml/Magento/backend/web/app/updater/styles/less'
    },
    uglify: {
        legacy: webroot + 'lib/web/legacy-build.min.js'
    },
    doc: webroot + 'lib/web/css/docs',
    spec: webroot + 'dev/tests/js/spec',
    static: {
        dir: webroot + 'dev/tests/static/testsuite/Magento/Test/Js/_files',
        whitelist: webroot + 'dev/tests/static/testsuite/Magento/Test/Js/_files/whitelist/',
        blacklist: webroot + 'dev/tests/static/testsuite/Magento/Test/Js/_files/blacklist/',
        tmp: webroot + 'validation-files.txt'
    },
    webroot: '../webroot/'
};
