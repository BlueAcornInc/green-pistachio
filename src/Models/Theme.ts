import glob from 'fast-glob';
import debug from 'debug';
const logger = debug('gpc:theme');

type ThemeConstructorArgs = ThemeData & {
    sourceDirectory: string;
    parent?: Theme;
};

// type CriticalCssInput = {
//     filepath: string;
//     urls: string[];
// };

export type ThemeData = {
    area: "frontend" | "adminhtml";
    path: string;
    enabled?: boolean;
    locales?: string[];
    stylesheets?: string[];
    // criticalCss?: CriticalCssInput[];
};

export default class Theme {
    private sourceDirectory: string;
    private parent?: Theme;
    private enabled: boolean | undefined;
    private data: ThemeData;

    constructor(config: ThemeConstructorArgs) {
        const {
            sourceDirectory,
            area,
            path,
            parent,
            locales,
            stylesheets,
            // criticalCss,
            enabled
        } = config;

        this.sourceDirectory = sourceDirectory;
        this.parent = parent;
        this.enabled = enabled || false;
        this.data = {
            area,
            path,
            locales,
            stylesheets,
            // criticalCss
        };
    }

    public configure(data: Partial<ThemeData>) {
        this.data = {
            ...this.data,
            ...data
        };
    }

    public getSourceDirectory() {
        return this.sourceDirectory;
    }

    public getParent() {
        return this.parent;
    }

    public getData() {
        return this.data;
    }

    public getLocales(): string[] {
        if (this.data.locales) {
            return this.data.locales;
        }

        logger(`${this.data.path}: Can't Determine Locales, defaulting to en_US`);

        return ['en_US'];
    }

    public getStyleSheets(): string[] {
        if (this.data.stylesheets) {
            return this.data.stylesheets;
        }

        logger(`${this.data.path}: Stylesheets not explicitly defined, detecting entries now.`);

        const files: Set<string> = new Set();

        const discoveredFiles = glob
            .sync(`${this.getSourceDirectory()}/**/[^_]*.less`)
            .map(sourceFile => sourceFile.replace(`${this.getSourceDirectory()}/web/`, ''));

        for (const file of discoveredFiles) {
            files.add(file);
        }

        let childTheme = this.getParent();

        while (childTheme) {
            for (const file of childTheme.getStyleSheets()) {
                files.add(file);
            }
            childTheme = childTheme.getParent();
        }

        const stylesheets = Array.from(files);

        logger(`${this.data.path}: Discovered the following stylesheets: ${stylesheets.join(', ')}`);

        return stylesheets;
    }

    // public getCriticalPaths(): CriticalCssInput[] {
    //     if (this.data.criticalCss) {
    //         return this.data.criticalCss;
    //     }

    //     logger(`No critical path data available for: ${this.data.path}`);

    //     return [];
    // }

    public getEnabled() {
        return this.enabled;
    }

    public setEnabled(enabled: undefined | boolean) {
        this.enabled = enabled;
    }
}
