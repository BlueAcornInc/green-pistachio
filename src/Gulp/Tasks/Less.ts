import {
    src,
    dest,
    parallel,
    watch,
    TaskFunction
} from 'gulp';
import debug from 'debug';
import sourcemaps from 'gulp-sourcemaps';
import less from 'gulp-less';
import livereload from 'gulp-livereload';
import { join } from 'path';
import Project from '../../Models/Project';
import Theme from '../../Models/Theme';
import { TaskInterface } from './TaskInterface';
const logger = debug('gpc:gulp:less');
import taskName from "./Decorators/TaskNameDecorator";

@taskName("less")
export default class Less implements TaskInterface {
    execute(project: Project, theme?: Theme) {
        const themes = theme ? [theme] : project.getThemes();
        let sources: string[] = [];

        for (const theme of themes) {
            for (const publicDirectory of project.getThemePubDirectories(theme)) {
                sources = sources.concat(theme.getStyleSheets().map(stylesheet => join(publicDirectory, stylesheet)));
            }
        }

        const tasks: TaskFunction[] = sources.map(source => {

            const task: TaskFunction = (done) => {
                logger(source);
                src(source)
                    .pipe(sourcemaps.init())
                    .pipe(
                        less()
                            .on('error', (err) => {
                                logger(source);
                                logger(err);
                            })
                    )
                    .pipe(sourcemaps.write('./'))
                    .pipe(
                        dest(
                            source.substring(
                                0,
                                source.lastIndexOf('/')
                            )
                        )
                    )
                    .pipe(livereload())
                    .on('finish', done);
            };

            task.displayName = `less<${source}>`;

            return task;
        });

        if (tasks.length === 0) {
            logger('No Less Tasks Running');
            tasks.push(done => done());
        }

        return parallel(...tasks);
    }

    watch(project: Project, theme?: Theme): TaskFunction {
        return (done) => {
            const themes = theme ? [theme] : project.getThemes();

            watch(
                this.getWatchFiles(project, themes),
                this.execute(project, theme)
            );
        };
    }

    private getWatchFiles(project: Project, themes: Theme[]): string[] {
        return project
            .getModules().map(module => `${module.getSourceDirectory()}/view/**/*.less`)
            .concat(
                themes.map(theme => `${theme.getSourceDirectory()}/**/*.less`)
            );
    }
}
