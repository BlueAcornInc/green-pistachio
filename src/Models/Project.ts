import { join } from 'path';
import { SyncHook } from 'tapable';
import debug from 'debug';
import Module from "./Module";
import Theme, { ThemeData } from "./Theme";
import ConfigLoader from './Project/ConfigLoader';
import { Configuration } from 'webpack';
import WebpackConfigFactory from '../Webpack/ConfigFactory';
const logger = debug('gpc:project');

type ProjectConstructorArgs = {
    root: string;
    themes: Theme[];
    modules: Module[];
    enabledThemes?: string[];
};

type WebpackDevServerConfig = {
    proxyUrl: string;
    devServerUrl: string;
    devServerPort: number;
};

export default class Project {
    public root: string;
    private themes: Theme[];
    private enabledThemes: Theme[];
    private modules: Module[];
    private webpackDevelopmentMode: boolean = false;
    private proxyUrl: string = '';
    private devServerConfig: Partial<WebpackDevServerConfig> = {};

    public hooks = {
        // These hooks allow for modifying various gulp configurations
        gulp: {
            babelConfig: new SyncHook(["config"]),
            babelTypeScriptConfig: new SyncHook(["config"]),
            pngSpriteConfig: new SyncHook(["config"]),
            svgSpriteConfig: new SyncHook(["config"]),
            eslintConfig: new SyncHook(["config"]),
            criticalCssConfig: new SyncHook(["config"]),
        },
        webpack: {
            config: new SyncHook(["config"]),
            entryGlobs: new SyncHook(["globs"]),
            appendConfig: new SyncHook<[Configuration[], WebpackConfigFactory]>(["webpackConfigs", "configFactory"])
        },
        configure: new SyncHook(["project"])
    };

    public experiments = {
        webpack: {
            // Set to true in the the project.configure hook to enable unstable HMR functionality
            hmr: false,
            // Set to true to enable css modules
            cssModules: false,
            // Set to true to emit files to the pub directory instead of the individual theme directories
            emitFilesToPub: false,
        }
    };

    constructor(config: ProjectConstructorArgs) {
        const { root, themes, modules, enabledThemes } = config;

        this.root = root;
        this.themes = themes;
        this.modules = modules;
        this.enabledThemes = enabledThemes
            ? themes.filter(theme => enabledThemes.includes(theme.getData().path))
            : [];
    }

    public configure() {
        if (this.enabledThemes.length > 0) {
            // Enable default themes
            for (const theme of this.enabledThemes) {
                theme.setEnabled(true);
            }
        } else {
            for (const theme of this.themes) {
                if (
                    !theme
                        .getSourceDirectory()
                        .includes(
                            join(
                                this.getRootDirectory(),
                                'vendor'
                            )
                        )
                ) {
                    theme.setEnabled(true);
                    this.enabledThemes.push(theme);
                }
            }
        }

        // Enable default modules
        for (const magentoModule of this.modules) {
            if (
                !magentoModule
                    .getSourceDirectory()
                    .includes(
                        join(
                            this.getRootDirectory(),
                            'vendor'
                        )
                    )
            ) {
                magentoModule.setEnabled(true);
            }
        }

        this.hooks.configure.call(this);
    }

    public configureTheme(themeData: Partial<ThemeData>) {
        if (!themeData.path) {
            logger(`can not configure theme without defined path: ${JSON.stringify(themeData)}`);
            return;
        }

        const theme = this.themes.find(theme => theme.getData().path === themeData.path);

        if (!theme) {
            logger(`can not find theme with path = '${themeData.path}'`);
            return;
        }

        theme.configure(themeData);

        if (themeData.hasOwnProperty('enabled')) {
            theme.setEnabled(themeData.enabled);
        }
    }

    public enableModule(moduleName: string) {
        const magentoModule = this.modules.find(magentoModule => magentoModule.getName() === moduleName);

        if (!magentoModule) {
            logger(`could not find module by name: ${moduleName}`);
            return;
        }

        magentoModule.setEnabled(true);
    }

    public getThemes() {
        return this.enabledThemes.filter(magentoTheme => magentoTheme.getEnabled());
    }

    public getAllThemes() {
        return this.themes;
    }

    public getModules() {
        return this.modules.filter(magentoModule => magentoModule.getEnabled());
    }

    public getAllModules() {
        return this.modules;
    }

    public getRootDirectory() {
        return this.root;
    }

    public getThemePubDirectories(theme: Theme) {
        return theme.getLocales().map(locale => join(
            this.getStaticDirectory(),
            theme.getData().area,
            theme.getData().path,
            locale
        ));
    }

    public getWebpackOutputDirectories(theme: Theme) {
        if (this.isWebpackDevelopmentMode() || this.experiments.webpack.emitFilesToPub) {
            return theme.getLocales().map(locale => join(
                this.getRootDirectory(),
                ...(
                    this.isWebpackDevelopmentMode()
                    ? ['static']
                    : ['pub', 'static']
                ),
                theme.getData().area,
                theme.getData().path,
                locale
            ));
        }

        return [join(theme.getSourceDirectory(), 'web')];
    }

    public getStaticDirectory() {
        return join(
            this.getRootDirectory(),
            'pub',
            'static'
        );
    }

    public isWebpackDevelopmentMode() {
        return this.webpackDevelopmentMode;
    }

    public setWebpackDevelopmentMode() {
        this.webpackDevelopmentMode = true;
    }

    public setProxyUrl(url: string) {
        this.proxyUrl = url;
    }

    public getProxyUrl(): string {
        if (this.proxyUrl.length === 0) {
            if (!debug.enabled('gpc:logger')) {
                debug.enable('gpc:logger');
            }
            logger(`A project URL must be provided in your projects ${ConfigLoader.CONFIG_FILE}`);
            logger(`\n\nExample:
module.exports = project => {
project.hooks.configure.tap("Set Product Url", () => {
    project.setProxyUrl("https://example.test");
});
}
            `);

            throw new Error(`A project URL must be provided in your projects ${ConfigLoader.CONFIG_FILE}`);
        }

        return this.proxyUrl;
    }

    public setWebpackDevServerConfig(config: Partial<WebpackDevServerConfig>) {
        if (config.proxyUrl) {
            this.proxyUrl = config.proxyUrl;
        }

        this.devServerConfig = config;
    }

    public getWebpackDevServerConfig(): WebpackDevServerConfig {
        return {
            proxyUrl: this.getProxyUrl(),
            devServerUrl: 'green-pistachio.test',
            devServerPort: 8080,
            ...(this.devServerConfig || {})
        };
    }
}
