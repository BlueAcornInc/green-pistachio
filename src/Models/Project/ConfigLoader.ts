import { promises as fs } from 'fs';
import { join } from 'path';
import debug from 'debug';
import Project from "../Project";
const logger = debug('gpc:configLoader');

export default class ConfigLoader {
    public static CONFIG_FILE = 'green-pistachio.config.js';

    constructor(private project: Project) {}

    get filePath(): string {
        return join(
            this.project.getRootDirectory(),
            ConfigLoader.CONFIG_FILE
        );
    }

    public async configExists(): Promise<boolean> {
        try {
            const configExists = await fs.stat(this.filePath);

            if (configExists) {
                return true;
            }
        } catch (err) {
            logger(`Config file can't be found at ${this.filePath}`);
        }

        return false;
    }

    public async loadConfig(): Promise<boolean> {
        const configExists = await this.configExists();

        if (configExists) {
            try {
                this.interopRequire()(this.project);
                return true;
            } catch (err) {
                logger(`Problem occurred while running user provided configuration file: ${err}`);
            }
        }

        return false;
    }

    private interopRequire() {
        const configModule = require(this.filePath);
        return configModule && configModule.__esModule ? configModule.default : configModule;
    }
}