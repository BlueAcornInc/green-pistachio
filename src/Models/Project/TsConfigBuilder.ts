import { promises as fs } from 'fs';
import { join } from 'path';
import debug from 'debug';
import Project from "../Project";
const logger = debug('gpc:typescript:configBuilder');

export default class TsConfigBuilder {
    public getFilePath(): string {
        return join(__dirname, '..', '..', '..', 'green-pistachio.tsconfig.json');
    }

    public async configFileExists(): Promise<boolean> {
        try {
            await fs.stat(this.getFilePath());

            return true;
        } catch (err) {
            return false;
        }
    }

    public async emitConfigFile(project: Project) {
        logger(`Generating Projects tsconfig.json base`);
        const include: string[] = [];
        const paths: Record<string, string[]> = {}

        for (const module of project.getModules()) {
            const moduleSrcDir = module.getSourceDirectory().replace(project.getRootDirectory(), '');

            const modulePaths: string[] = [
                join(moduleSrcDir, 'view/frontend/web/ts/*'),
                join(moduleSrcDir, 'view/base/web/ts/*'),
                join(moduleSrcDir, 'view/adminhtml/web/ts/*'),
            ];

            paths[`${module.getName()}/*`] = modulePaths.map(modulePath => modulePath.replace(/^\//, ''));

            include.push(join(moduleSrcDir, 'view/frontend/web/ts/**/*'));
            include.push(join(moduleSrcDir, 'view/base/web/ts/**/*'));
            include.push(join(moduleSrcDir, 'view/adminhtml/web/ts/**/*'));
        }

        for (const theme of project.getAllThemes()) {
            const themeSrcDir = theme.getSourceDirectory().replace(project.getRootDirectory(), '');

            include.push(join(themeSrcDir, '**/*'));
            
            // Add theme overrides to module paths
            for (const moduleName of Object.keys(paths)) {
                paths[moduleName].unshift(
                    join(themeSrcDir, moduleName.replace(/\/\*$/, ''), 'web/ts/*').replace(/^\//, '')
                );
            }
        }

        logger(`Resolved include paths to: ${JSON.stringify(include)}`);
        logger(`Resolved module paths to: ${JSON.stringify(paths)}`);

        const writePath = this.getFilePath();
        const config = this.getBaseConfig();

        config.include = include.map(includePath => `./${includePath}`);
        config.compilerOptions.paths = paths;

        try {
            logger(`Writing config file: ${writePath}`);
            await fs.writeFile(writePath, JSON.stringify(config, null, 4));
        } catch (err) {
            logger(`Problem writing config file: ${writePath}`);
            throw err;
        }

        return writePath;
    }

    private getBaseConfig() {
        return {
            "include": [
              "./app/**/*"
            ],
            "compilerOptions": {
              "baseUrl": ".",
              "paths": {},
              "strict": true,
              "esModuleInterop": true,
              "lib": [
                "dom",
                "es2015"
              ],
              "jsx": "react",
              "emitDecoratorMetadata": true,
              "experimentalDecorators": true,
              "target": "es5",
              "allowJs": true,
              "skipLibCheck": true,
              "allowSyntheticDefaultImports": true,
              "forceConsistentCasingInFileNames": true,
              "module": "esnext",
              "moduleResolution": "node",
              "resolveJsonModule": true,
              "isolatedModules": true,
              "noEmit": true
            }
          }
    }
}