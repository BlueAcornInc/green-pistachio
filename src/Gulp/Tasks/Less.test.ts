/**
 * @jest-environment node
 */
import mock from 'mock-fs';
import { join } from 'path';
import Project from '../../Models/Project';
import Theme from '../../Models/Theme';
import Less from './Less';
import { promises as fs } from 'fs';

afterEach(() => {
    mock.restore();
});

describe('Gulp: Less', () => {
    it('should compile less', ()=> {});

    //TODO: Rewrite Less Tests
    // it('should compile less', async (done) => {
    //     mock({
    //         'node_modules': mock.load(join(__dirname, '..', '..', '..', 'node_modules')),
    //         'vendor/magento/theme-frontend-blank/web/css/styles.less': '',
    //         'pub/static/frontend/Magento/blank/en_US/css/styles.less': '@color: blue; p { color: @color }'
    //     });

    //     const project = new Project({
    //         themes: [
    //             new Theme({
    //                 sourceDirectory: `vendor/magento/theme-frontend-blank`,
    //                 area: 'frontend',
    //                 path: 'Magento/blank',
    //                 enabled: true,
    //             })
    //         ],
    //         modules: [],
    //         root: '',
    //     });

    //     const less = new Less();
    //     less.execute(project)(async () => {
    //         const styles = await fs.readFile(`pub/static/frontend/Magento/blank/en_US/css/styles.css`);
    //         expect(styles.toString()).toMatchSnapshot();
    //         done();
    //     });
    // });
});
