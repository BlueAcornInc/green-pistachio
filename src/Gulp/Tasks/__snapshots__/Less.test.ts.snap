// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gulp: Less should compile less 1`] = `
"p {
  color: blue;
}

/*# sourceMappingURL=styles.css.map */
"
`;
